copyright_header/copyright_header.py

```diff
@@ -0,0 +1,221 @@
+    """
+    Parse a copyright year range from a header string.
+
+    Looks for a year range of the form "2019" or "2010-2019", and
+    returns the start and end year.
+
+    If there are multiple year ranges, parses only the first.
+
+    Parameters
+    ----------
+    header_text : str
+        The text to be parsed. Could be the entire copyright header,
+        or a single line from the copyright header.
+
+    Returns
+    -------
+    start_year, end_year : int
+        Start year and end year described by the range.
+    match_pos : int
+        Position within the string at which the match occurred.
+
+    Raises
+    ------
+    ValueError
+        If no year range is recognised from the given string.
+    """
+    years_match = re.search(YEAR_RANGE, header_text)
+    if not years_match:
+        raise ValueError("No year range found in the given string.")
+
+    start_year = int(years_match.group("start_year"))
+    end_year_str = years_match.group("end_year")
+    end_year = int(end_year_str) if end_year_str is not None else start_year
+    return start_year, end_year, years_match.start()
+
+
+class HeaderError:
+    """
+    Base class for the copyright header errors.
+    """
+    def __init__(self, lineno, col_offset):
+        self.lineno = lineno
+        self.col_offset = col_offset
+
+    @property
+    def full_message(self):
+        """
+        Full message in the form expected by flake8 (including the error code).
+        """
+        return "{} {}".format(self.code, self.message)
+
+
+class MissingCopyrightHeaderError(HeaderError):
+    """
+    Error reported when no copyright header can be identified.
+    """
+    code = "H101"
+    message = "Missing copyright header"
+
+
+class DuplicateCopyrightHeaderError(HeaderError):
+    """
+    Error reported if multiple copyright headers found.
+    """
+    code = "H102"
+    message = "Multiple copyright headers found"
+
+
+class IncorrectCopyrightHeaderError(HeaderError):
+    """
+    Error reported if a copyright header is found, but its wording
+    doesn't match the officially approved wording.
+    """
+    code = "H103"
+    message = "Wrong copyright header found"
+
+
+class OutdatedCopyrightYearError(HeaderError):
+    """
+    Error reported if the copyright header doesn't have the correct
+    year information in it.
+    """
+    code = "H104"
+
+    def __init__(self, lineno, col_offset, end_year):
+        super().__init__(lineno=lineno, col_offset=col_offset)
+        self.end_year = end_year
+
+    @property
+    def message(self):
+        return (
+            "Copyright end year ({}) out of date. The year should be at "
+            "least {}.".format(self.end_year, MINIMUM_END_YEAR)
+        )
+
+
+def copyright_header(filename):
+    """
+    Check copyright header presence and accuracy in a Python file.
+    """
+    # Use tokenize.open to ensure that we respect the coding cookie.
+    with tokenize.open(filename) as f:
```

Comment:
Flake8 can also read its input from stdin (e.g. when you run `flake8 - < myfile.py`) and some IDEs use this to call flake8. Would it be possible to address that case as well? E.g. 

```python
from flake8 import utils

def copyright_header(filename):
    if filename == "stdin":
        # Flake8 allows for input to be provided via stdin (e.g. via ``flake8 -
        # < my_file.py`` or by piping in the output of another command). In
        # this case, ``filename`` is set to ``stdin`` and we have to read the
        # input directly from stdin. Many editors pass in input via stdin.
        #
        # Note: this will break if the user sets a different name for stdin via
        # the --stdin-display-name option for flake8.
        file_contents = utils.stdin_get_value()
    else:
        # ...
```

Reply:
Will do. Is there a better way to check for this case? It bugs me to special-case something that's actually a valid filename, even though it's highly unlikely to be the name of an actual file.

Reply:
And it appears that the actual filename received is configurable:
```
$ cat setup.py | flake8 --stdin-display-name=bob -
Traceback (most recent call last):
  File "/Users/mdickinson/.venvs/traits/bin/flake8", line 10, in <module>
    sys.exit(main())
[... more traceback lines omitted ...]
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/tokenize.py", line 392, in open
    buffer = _builtin_open(filename, 'rb')
FileNotFoundError: [Errno 2] No such file or directory: 'bob'
```

Reply:
Okay, I think I found a way. Will update shortly.

Reply:
I actually looked into this in another context, and to my knowledge there is no easy way of getting the `stdin_display_name` value from the environment (that's what the comment was referring to). 

I looked at a few other flake8 plugins (e.g. flake8_black) but none of them had a reasonably sane way of doing this.

A way of bypassing the issue with `stdin` being potentially a valid filename would be to check if stdin is a valid tty. Not sure what you think about that?
```
$ python -c "import os,sys; print(os.isatty(sys.stdin.fileno()))"
True
$ echo bar | python -c "import os,sys; print(os.isatty(sys.stdin.fileno()))"
False
``

Reply:
Sorry but is there something missing in the last commit? With the latest version I still see the error:
```
$ flake8 - < setup.py 
Traceback (most recent call last):
  File "/home/jvankerschaver/.edm/envs/pyface-test/bin/flake8", line 11, in <module>
    sys.exit(main())
  File "/home/jvankerschaver/.edm/envs/pyface-test/lib/python3.6/site-packages/flake8/main/cli.py", line 18, in main
    app.run(argv)
  File "/home/jvankerschaver/.edm/envs/pyface-test/lib/python3.6/site-packages/flake8/main/application.py", line 393, in run
    self._run(argv)
  File "/home/jvankerschaver/.edm/envs/pyface-test/lib/python3.6/site-packages/flake8/main/application.py", line 381, in _run
    self.run_checks()
  File "/home/jvankerschaver/.edm/envs/pyface-test/lib/python3.6/site-packages/flake8/main/application.py", line 300, in run_checks
    self.file_checker_manager.run()
  File "/home/jvankerschaver/.edm/envs/pyface-test/lib/python3.6/site-packages/flake8/checker.py", line 331, in run
    self.run_serial()
  File "/home/jvankerschaver/.edm/envs/pyface-test/lib/python3.6/site-packages/flake8/checker.py", line 315, in run_serial
    checker.run_checks()
  File "/home/jvankerschaver/.edm/envs/pyface-test/lib/python3.6/site-packages/flake8/checker.py", line 598, in run_checks
    self.run_ast_checks()
  File "/home/jvankerschaver/.edm/envs/pyface-test/lib/python3.6/site-packages/flake8/checker.py", line 502, in run_ast_checks
    for (line_number, offset, text, check) in runner:
  File "/home/jvankerschaver/src/enthought/ets/traits/copyright_header/copyright_header.py", line 215, in run
    for error in copyright_header(self.filename):
  File "/home/jvankerschaver/src/enthought/ets/traits/copyright_header/copyright_header.py", line 151, in copyright_header
    with tokenize.open(filename) as f:
  File "/home/jvankerschaver/.edm/envs/pyface-test/lib/python3.6/tokenize.py", line 452, in open
    buffer = _builtin_open(filename, 'rb')
FileNotFoundError: [Errno 2] No such file or directory: 'stdin'
```

Reply:
I don't think so. Did you re-run the `pip install` for `copyright_header`?

Reply:
Sorry, yes, something's wrong. Not sure what's happened. Will fix.

Reply:
Apologies; I messed up (editing the installed copy instead of the repo copy) I think it should be okay now. Please could you double check?

Reply:
Works splendidly, thanks for fixing!

```
$ flake8 --stdin-display-name joseph - < setup.py 
joseph:1:1: H103 Wrong copyright header found
joseph:36:1: H102 Multiple copyright headers found
joseph:164:1: E302 expected 2 blank lines, found 1
```